crates/libs/registry/src/key.rs

```diff
@@ -0,0 +1,327 @@
+    pub fn create<T: AsRef<str>>(&self, path: T) -> Result<Self> {
+        let mut handle = 0;
+
+        let result = unsafe {
+            RegCreateKeyExW(
+                self.0,
+                pcwstr(path).as_ptr(),
+                0,
+                std::ptr::null(),
+                REG_OPTION_NON_VOLATILE,
+                KEY_READ | KEY_WRITE,
+                std::ptr::null(),
+                &mut handle,
+                std::ptr::null_mut(),
+            )
+        };
+
+        win32_error(result).map(|_| Self(handle))
+    }
+
+    /// Opens a registry key.
+    pub fn open<T: AsRef<str>>(&self, path: T) -> Result<Self> {
+        let mut handle = 0;
+
+        let result =
+            unsafe { RegOpenKeyExW(self.0, pcwstr(path).as_ptr(), 0, KEY_READ, &mut handle) };
+
+        win32_error(result).map(|_| Self(handle))
+    }
+
+    /// Removes the registry keys and values of the specified key recursively.
+    pub fn remove_tree<T: AsRef<str>>(&self, path: T) -> Result<()> {
+        let result = unsafe { RegDeleteTreeW(self.0, pcwstr(path).as_ptr()) };
+        win32_error(result)
+    }
+
+    /// Removes the registry value.
+    pub fn remove_value<T: AsRef<str>>(&self, name: T) -> Result<()> {
+        let result = unsafe { RegDeleteValueW(self.0, pcwstr(name).as_ptr()) };
+        win32_error(result)
+    }
+
+    /// Creates an iterator of registry key names.
+    pub fn keys(&self) -> Result<KeyIterator<'_>> {
+        KeyIterator::new(self)
+    }
+
+    /// Creates an iterator of registry values.
+    pub fn values(&self) -> Result<ValueIterator<'_>> {
+        ValueIterator::new(self)
+    }
+
+    /// Sets the name and value in the registry key.
+    pub fn set_u32<T: AsRef<str>>(&self, name: T, value: u32) -> Result<()> {
+        let result = unsafe {
+            RegSetValueExW(
+                self.0,
+                pcwstr(name).as_ptr(),
+                0,
+                REG_DWORD,
+                &value as *const _ as _,
+                4,
+            )
+        };
+
+        win32_error(result)
+    }
+
+    /// Sets the name and value in the registry key.
+    pub fn set_u64<T: AsRef<str>>(&self, name: T, value: u64) -> Result<()> {
+        let result = unsafe {
+            RegSetValueExW(
+                self.0,
+                pcwstr(name).as_ptr(),
+                0,
+                REG_QWORD,
+                &value as *const _ as _,
+                8,
+            )
+        };
+
+        win32_error(result)
+    }
+
+    /// Sets the name and value in the registry key.
+    pub fn set_string<T: AsRef<str>>(&self, name: T, value: T) -> Result<()> {
+        let value = pcwstr(value);
+
+        let result = unsafe {
+            RegSetValueExW(
+                self.0,
+                pcwstr(name).as_ptr(),
+                0,
+                REG_SZ,
+                value.as_ptr() as _,
+                value.len() as u32 * 2,
```

Comment:
While the length overflowing a `u32` may be exceedingly rare (and probably indicates something else has gone badly wrong) I do think it'd be good to catch that here rather than silently truncating. Maybe it can return an error? Ditto the other similar functions.

Reply:
How about this: [e8a892e](https://github.com/microsoft/windows-rs/pull/2848/commits/e8a892e6d15f56a889842c5c3a8907056f6479a9)

Reply:
Sure, that works for me!